src/vs/workbench/contrib/notebook/browser/notebookExecutionStateServiceImpl.ts

```diff
@@ -140,51 +140,68 @@ export class NotebookExecutionStateService extends Disposable implements INotebo
 		this._cellListeners.set(CellUri.generate(notebookUri, cellHandle), disposable);
 
-		notebook.onWillAddRemoveCells((e: NotebookTextModelWillAddRemoveEvent) => {
-			const lastFailedCell = this.getLastFailedCellForNotebook(notebook.uri);
-			e.rawEvent.changes.forEach(([start, deleteCount]) => {
-				if (deleteCount) {
-					const deletedHandles = notebook.cells.slice(start, start + deleteCount).map(c => c.handle);
-					if (lastFailedCell && deletedHandles.includes(lastFailedCell)) {
-						this._setLastFailedCellVisibility(notebook.uri, false);
-					}
-				}
-			});
-		});
+		return exe;
+	}
 
-		notebook.onDidChangeContent((events) => {
-			const lastFailedCell = this.getLastFailedCellForNotebook(notebook.uri);
-			events.rawEvents.forEach((e) => {
-				if (e.kind === NotebookCellsChangeType.ModelChange) {
-					e.changes.forEach(([start, deleteCount]) => {
-						if (deleteCount === 0) {
-							const addHandle = notebook.cells[start].handle;
-							if (addHandle === lastFailedCell) {
-								this._setLastFailedCellVisibility(notebook.uri, true);
-							}
-						}
-					});
-				}
-			});
-		});
+	private _setLastFailedCell(notebookURI: URI, cellHandle: number) {
+		this._lastFailedCells.set(notebookURI, cellHandle);
 
-		return exe;
+		if (this._lastFailedCellDisposableListeners.get(notebookURI) === undefined) {
+			const notebook = this._notebookService.getNotebookTextModel(notebookURI);
+			if (notebook) {
+				this._setFailedCellListeners(notebook);
+			}
+		}
+		this._onDidChangeLastRunFailState.fire({ failed: true, notebook: notebookURI });
 	}
 
-	private _setLastFailedCellVisibility(notebook: URI, visible: boolean) {
-		this._lastFailedCellsVisible.set(notebook, visible);
-		this._onDidChangeLastRunFailState.fire({ failed: visible, notebook });
+	private _clearLastFailedCell(notebookURI: URI) {
+		this._lastFailedCells.delete(notebookURI);
+		this._removeFailedCellListeners(notebookURI);
+		this._onDidChangeLastRunFailState.fire({ failed: false, notebook: notebookURI });
 	}
 
-	private _setLastFailedCell(notebook: URI, cellHandle: number) {
-		this._lastFailedCells.set(notebook, cellHandle);
-		this._lastFailedCellsVisible.set(notebook, true);
-		this._onDidChangeLastRunFailState.fire({ failed: true, notebook });
+	private _setFailedCellListeners(notebook: NotebookTextModel) {
+		const lastFailedCellDisposableListener = combinedDisposable(
+			notebook.onWillAddRemoveCells((e: NotebookTextModelWillAddRemoveEvent) => {
+				const lastFailedCell = this.getLastFailedCellForNotebook(notebook.uri);
+				e.rawEvent.changes.some(([start, deleteCount]) => {
+					if (deleteCount) {
+						const lastFailedCellPos = notebook.cells.findIndex(c => c.handle === lastFailedCell);
+						if (lastFailedCellPos >= start && lastFailedCellPos < start + deleteCount) {
+							this._onDidChangeLastRunFailState.fire({ failed: false, notebook: notebook.uri });
```

Comment:
Seems tricky that this will fire but `getLastFailedCellForNotebook` will still return that cell. That method should agree with this event.